File: python/LevelOrder102.py
import collections
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # def levelOrder(self, root: TreeNode) -> List[List[int]]:
    #     if not root:
    #         return []
    #     queue = collections.deque()
    #     result = []
    #     queue.append(root)
    #     while queue:
    #         length = len(queue)
    #         list = []
    #         for _ in range(0, length):
    #             node = queue.popleft()
    #             list.append(node.val)
    #             if node.left:
    #                 queue.append(node.left)
    #             if node.right:
    #                 queue.append(node.right)
    #         result.append(list)
    #     return result
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root: return []
        self.result = []
        self._dfs(root, 0)
        return self.result
    
    def _dfs(self, node, level):
        if not node:
            return
        if len(self.result) < level + 1:
            self.result.append([])
        self.result[level].append(node.val)
        self._dfs(node.left, level+1)
        self._dfs(node.right, level+1)

File: java/LevelOrder102.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    // public List<List<Integer>> levelOrder(TreeNode root) {
    //     List<List<Integer>> result = new ArrayList<>();
    //     if(root == null)
    //         return result;
    //     Queue<TreeNode> queue = new LinkedList<>();
    //     queue.add(root);
    //     while(!queue.isEmpty()){
    //         int size = queue.size();
    //         List<Integer> list = new ArrayList<>();
    //         while(size > 0){
    //             TreeNode node = queue.poll();
    //             list.add(node.val);
    //             if(node.left != null) queue.add(node.left);
    //             if(node.right != null) queue.add(node.right);
    //             size--;
    //         }
    //         result.add(list);
    //     }
    //     return result;
    // }
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if(root == null)
            return new ArrayList<>();
        DFS(root, 0, result);
        return result;
    }
    public void DFS(TreeNode node, int level, List<List<Integer>> result) {
        if(node == null)
            return;
        if(result.size() < level + 1)
            result.add(new ArrayList<>());
        result.get(level).add(node.val);
        if(node.left != null) DFS(node.left, level + 1, result);
        if(node.right != null) DFS(node.right, level + 1,result);
    }
}
